cEP0002.md

```diff
@@ -0,0 +1,350 @@
+3. Each bear will generate the following type of tasks:
+    * Global bears fire tasks that process all files.
+    * Local bears fire tasks for each individual file.
+    * Arbitrary tasks such as dependency bears or other tasks are also
+    possible.
+4. A bear that has dependencies will only be scheduled to the process pool once all its dependencies have been resolved. It's done in the following way:
+    * It's checked if the bear has some dependencies.
+    * If they haven't been resolved the bear is not scheduled.
+    * As soon as the dependencies are resolved for a bear it is removed from the dependency dictionary and scheduled to the `ProcessPoolExecutor`.
+5. A special purpose `Dependencies` class will be introduced to do the above
+mentioned task. This will be responsible for keeping track of dependencies.
+Whenever a bear completes it will send a signal so that dependencies can be
+resolved.
+
+# The File Proxy
+
+With the new bear design we are also moving to a new wrapper object to wrap
+every file. This object will also contain some metadata such as the filename
+of the file. The file proxy is an object that contains properties that are
+cached and passed around processes.
+
+What it will cache:
+    * The filename, this is the starting point for every other attribute.
+    * The content of the file as a string.
+    * The content of the file, line-splitted
+    * It can also cache other things about files as and when required.
+
+The file proxy tries to load from file using the filename if no other cached
+property is already loaded.
+
+A prototype of the file proxy:
+
+```python
+class FileProxy:
+    def __init__(self, filename):
+        self.filename = filename
+        self.lines = []
+        self.content = ""
+        self.load()
+
+    def __iter__(self):
+        return iter(self.lines)
+
+    def load(self):
+        with open(self.filename) as filehandle:
+            self.content = filehandle.read()
+            self.lines = self.content.splitlines(True)
+
+    def __eq__(self, other):
+        if self.filename == other.filename:
+            return True
+        return False
+
+    def __hash__(self):
+        return hash(self.filename)
+```
+
+# Dependency Tracking
+
+A new class will be responsible for keeping track of the dependencies of different bear and it will work in the following manner:
+    * Each bear will add it's dependencies to this tracker class.
+    * Whenever a bear completes it's execution the `resolve()` method of the
+    new class will be called with this bear as an argument.
+    * This would execute the marking of this bear as complete in the
+    dependency tracker.
+    * Once the marking is complete the tracker returns all bears which have no
+    dependencies remaining as pending, so that their tasks can be scheduled to
+    the process pool.
+
+Here is a prototype:
+
+```python
+from collections import defaultdict
+
+
+class Dependencies:
```

Comment:
can you add documentation comments? At first this was confusing but actually this new dependency resolving is super beautiful!